drivers/st25r3911b/st25r3911b_interrupt.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <kernel.h>
+#include <device.h>
+#include <soc.h>
+#include <gpio.h>
+#include <logging/log.h>
+#include <drivers/st25r3911b_spi.h>
+#include <drivers/st25r3911b_interrupt.h>
+
+LOG_MODULE_DECLARE(st25r3911b);
+
+#define IRQ_REG_CNT 3
+
+static struct gpio_callback gpio_cb;
+static struct device *gpio_dev;
+
+static atomic_t irq_mask;
+static struct k_sem *sem;
+
+
+static void irq_pin_cb(struct device *gpiob, struct gpio_callback *cb,
+		       u32_t pins)
+{
+	k_sem_give(sem);
+}
+
+static int gpio_init(void)
+{
+	int err;
+
+	gpio_dev = device_get_binding(DT_GPIO_P0_DEV_NAME);
+	if (!gpio_dev) {
+		LOG_ERR("GPIO device binding error.");
+		return -ENXIO;
+	}
+
+	/* Configure IRQ pin */
+	err = gpio_pin_configure(gpio_dev, CONFIG_ST25R3911B_IRQ_PIN,
+				 GPIO_DIR_IN | GPIO_INT |
+				 GPIO_INT_EDGE | GPIO_INT_ACTIVE_HIGH);
+	if (err) {
+		return err;
+	}
+
+	gpio_init_callback(&gpio_cb, irq_pin_cb,
+			   BIT(CONFIG_ST25R3911B_IRQ_PIN));
+
+	err = gpio_add_callback(gpio_dev, &gpio_cb);
+	if (err) {
+		return err;
+	};
+
+	return gpio_pin_enable_callback(gpio_dev, CONFIG_ST25R3911B_IRQ_PIN);
+}
+
+u32_t st25r3911b_irq_read(void)
+{
+	u8_t data[IRQ_REG_CNT] = {0};
+	u32_t status = 0;
+	u32_t pin = 0;
+	int err;
+
+	/* In some cases, interrupts come so quickly that
+	 * the state of the pin will not change
+	 */
+	do {
+
+		err = st25r3911b_multiple_reg_read(ST25R3911B_REG_MAIN_INT,
+						   data, ARRAY_SIZE(data));
+		__ASSERT_NO_MSG(!err);
+
+		status = (u32_t)data[0];
+		status |= ((u32_t)data[1] << 8);
+		status |= ((u32_t)data[2] << 16);
+
+		gpio_pin_read(gpio_dev, CONFIG_ST25R3911B_IRQ_PIN, &pin);
+
+	} while (pin);
+
+	LOG_DBG("Read interrupt, status: %u", status);
+
+	status &= ~atomic_get(&irq_mask);
+
+	return status;
+}
+
+int st25r3911b_irq_modify(u32_t clr_mask, u32_t set_mask)
+{
+	int err;
+	u32_t mask;
+	u32_t old_mask;
+
+	old_mask = atomic_get(&irq_mask);
+	mask = (~old_mask & set_mask) | (old_mask & clr_mask);
```

Comment:
I don't like this... `irq_mask` can potentially be modified while we work on a local copy. Would feel better if you spinlocked this part for modification.

Reply:
Using this atomic get and set here would work as expected. Spinlock should be used here.

Reply:
Ok, Changed to spinlock.